src/System.CommandLine/Parsing/ParseResultVisitor.cs

```diff
@@ -483,22 +483,31 @@ private void ValidateAndConvertOptionResult(OptionResult optionResult)
 
         private void ValidateAndConvertArgumentResult(ArgumentResult argumentResult)
         {
-            if (argumentResult.Argument is { } argument)
-            {
-                var parseError =
-                    argumentResult.Parent?.UnrecognizedArgumentError(argument) ??
-                    argumentResult.CustomError(argument);
+            var argument = argumentResult.Argument;
 
-                if (parseError is { })
-                {
-                    AddErrorToResult(argumentResult, parseError);
-                    return;
-                }
+            var parseError =
+                argumentResult.Parent?.UnrecognizedArgumentError(argument) ??
+                argumentResult.CustomError(argument);
+
+            if (parseError is { })
+            {
+                AddErrorToResult(argumentResult, parseError);
+                return;
             }
 
-            if (argumentResult.GetArgumentConversionResult() is FailedArgumentConversionResult failed 
+            if (argumentResult.GetArgumentConversionResult() is FailedArgumentConversionResult failed
                 and not FailedArgumentConversionArityResult)
             {
+                if (argument.Parents.FirstOrDefault() is Option option)
+                {
+                    var completions = option.GetCompletions().ToArray();
```

Comment:
This is a way better fix than the enum-specific one I suggested on the issue. Nice reuse of existing domain concepts 👍